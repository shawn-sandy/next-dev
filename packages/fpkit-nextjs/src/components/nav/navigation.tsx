import React from "react";
import { Tag, Title } from "@fpkit/react";
import Nav, { NavItem } from "./nav";

import { MdxFile, Folder, PageMapItem } from "nextra";

type Kind = "Meta" | "MdxPage" | "Folder";

interface Meta {
  index: string;
  another: string;
  advanced: string;
  about: {
    title: string;
    type: string;
  };
  contact: {
    title: string;
    type: string;
    href: string;
    newWindow: boolean;
  };
  kind: "Meta";
}

type Data = (Meta | MdxFile | Folder | PageMapItem) & {
  kind: Kind;
};

function getFolders(data: Data[]): Folder[] {
  const folders: Folder[] = [];

  data.forEach((obj) => {
    if (obj.kind === "Folder") {
      folders.push(obj);
    }
  });

  return folders;
}

export interface Brand {
  brand?: {
    url?: string;
    logo?: React.ReactNode;
  };
}

interface NavbarProps extends Brand {
  styles?: {};
  data: Data[];
}

export const Navigation: React.FC<NavbarProps> = ({ data, brand, styles }) => {
  const folders = getFolders(data);

  return (
    <Nav styles={styles}>
      {!!brand && (
        <Tag as="div">
          <Title elm="h1" styles={{ "--fs": "1.2rem" }}>
            <Tag
              as="a"
              href={brand.url}
              styles={{ textTransform: "capitalize" }}
            >
              <>{brand.logo}</>
            </Tag>
          </Title>
        </Tag>
      )}

      <Tag as="ul">
        {folders.map((folder) => (
          <NavItem route={folder.route} key={folder.name}>
            <>{folder.name}</>
          </NavItem>
        ))}
      </Tag>
    </Nav>
  );
};

export default Navigation;
