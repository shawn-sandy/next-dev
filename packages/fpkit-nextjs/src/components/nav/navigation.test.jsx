import React from "react";

import { render, screen } from "@testing-library/react";
import { axe } from "vitest-axe";
import { expect, describe, test } from "vitest";

import { dataSet, branding } from "../data/navigation.mocks";

// ðŸ‘‰ import your component here
import Navigation from "./navigation";

describe("Navigation", () => {
  test("renders the navigation", () => {
    render(<Navigation data={dataSet} />);
    const navigation = screen.getByRole("navigation");

    expect(navigation).toBeDefined();
    expect(navigation).toMatchSnapshot();
  });

  test("is accessible", async () => {
    const { container } = render(<Navigation data={dataSet} />);

    expect(await axe(container)).toHaveNoViolations();
  });

  test("renders the branding w/navigation", async () => {
    render(<Navigation data={dataSet} brand={branding} />);
    const navigation = screen.getByRole("navigation");
    const brand = screen.getByRole("heading");

    expect(brand).toBeDefined();
    expect(brand).toMatchSnapshot();
    // @ts-ignore
    expect(await axe(navigation)).toHaveNoViolations();
    expect(navigation).toMatchSnapshot();
  });
});
