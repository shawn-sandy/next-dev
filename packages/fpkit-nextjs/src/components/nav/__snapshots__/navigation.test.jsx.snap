// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > renders the branding w/navigation 1`] = `
<h1
  style="--fs: 1.2rem;"
>
  <a
    href="/"
    style="text-transform: capitalize;"
  >
    Branding Logo
  </a>
</h1>
`;

exports[`Navigation > renders the branding w/navigation 2`] = `
<nav>
  <div>
    <h1
      style="--fs: 1.2rem;"
    >
      <a
        href="/"
        style="text-transform: capitalize;"
      >
        Branding Logo
      </a>
    </h1>
  </div>
  <ul>
    <li>
      <a
        href="/advanced"
        style="text-transform: capitalize;"
      >
        advanced
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Navigation > renders the navigation 1`] = `
<nav>
  <ul>
    <li>
      <a
        href="/advanced"
        style="text-transform: capitalize;"
      >
        advanced
      </a>
    </li>
  </ul>
</nav>
`;
